src/backend/CommandBufferStateTracker.cpp

```diff
@@ -15,69 +15,123 @@
-            DAWN_RETURN_ERROR("No active compute pipeline");
+        // Generate an error immediatly if a non-lazy aspect is missing as computing lazy aspects
+        // requires the pipeline to be set.
+        if ((missingAspects & ~kLazyAspects).any()) {
+            return GenerateAspectError(missingAspects);
         }
-        // Compute the lazily computed mAspects
-        if (!RecomputeHaveAspectBindGroups()) {
-            DAWN_RETURN_ERROR("Bind group state not valid");
+
+        RecomputeLazyAspects(missingAspects);
+
+        missingAspects = requiredAspects & ~mAspects;
+        if (missingAspects.any()) {
+            return GenerateAspectError(missingAspects);
         }
+
         return {};
     }
 
-    MaybeError CommandBufferStateTracker::ValidateCanDrawArrays() {
-        constexpr ValidationAspects requiredAspects = 1 << VALIDATION_ASPECT_PIPELINE |
-                                                      1 << VALIDATION_ASPECT_BIND_GROUPS |
-                                                      1 << VALIDATION_ASPECT_VERTEX_BUFFERS;
-        if ((requiredAspects & ~mAspects).none()) {
-            // Fast return-true path if everything is good
-            return {};
+    void CommandBufferStateTracker::RecomputeLazyAspects(ValidationAspects aspects) {
+        ASSERT(mAspects[VALIDATION_ASPECT_PIPELINE]);
+        ASSERT((aspects & ~kLazyAspects).none());
+
+        if (aspects[VALIDATION_ASPECT_BIND_GROUPS]) {
+            bool matches = true;
+
+            for (uint32_t i : IterateBitSet(mLastPipelineLayout->GetBindGroupLayoutsMask())) {
+                if (mLastPipelineLayout->GetBindGroupLayout(i) != mBindgroups[i]->GetLayout()) {
+                    matches = false;
+                }
+            }
+
+            if (matches) {
+                mAspects.set(VALIDATION_ASPECT_BIND_GROUPS);
+            }
         }
 
-        return RevalidateCanDraw();
+        if (aspects[VALIDATION_ASPECT_VERTEX_BUFFERS]) {
+            ASSERT(mLastRenderPipeline != nullptr);
+
+            auto requiredInputs = mLastRenderPipeline->GetInputState()->GetInputsSetMask();
+            if ((mInputsSet & requiredInputs) == requiredInputs) {
+                mAspects.set(VALIDATION_ASPECT_VERTEX_BUFFERS);
+            }
+        }
     }
 
-    MaybeError CommandBufferStateTracker::ValidateCanDrawElements() {
-        constexpr ValidationAspects requiredAspects =
-            1 << VALIDATION_ASPECT_PIPELINE | 1 << VALIDATION_ASPECT_BIND_GROUPS |
-            1 << VALIDATION_ASPECT_VERTEX_BUFFERS | 1 << VALIDATION_ASPECT_INDEX_BUFFER;
-        if ((requiredAspects & ~mAspects).none()) {
-            // Fast return-true path if everything is good
-            return {};
+    MaybeError CommandBufferStateTracker::GenerateAspectError(ValidationAspects aspects) {
+        ASSERT(aspects.any());
+
+        if (aspects[VALIDATION_ASPECT_INDEX_BUFFER]) {
+            DAWN_RETURN_ERROR("Missing index buffer");
```

Comment:
Is it the case that only the first failing aspect will be displayed?

Reply:
Yes, we'll want to have a making great error messages but that would be between the MVP and v1 I think.

Reply:
Acknowledged.